internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+package nwcth
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"sync"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/ooni/probe-cli/v3/internal/engine/experiment/nwebconnectivity"
+)
+
+type (
+	// CtrlRequest is the request sent to the test helper
+	CtrlRequest = nwebconnectivity.ControlRequest
+
+	// CtrlResponse is the response from the test helper
+	CtrlResponse = nwebconnectivity.ControlResponse
+
+	CtrlURLMeasurement = nwebconnectivity.ControlURL
+
+	CtrlEndpointMeasurement = nwebconnectivity.ControlEndpoint
+
+	CtrlHTTPMeasurement = nwebconnectivity.ControlHTTP
+
+	CtrlH3Measurement = nwebconnectivity.ControlH3
+
+	CtrlTLSMeasurement = nwebconnectivity.ControlTLSHandshake
+
+	CtrlHTTPRequest = nwebconnectivity.ControlHTTPRequest
+)
+
+// NextLocationInfo contains the redirected location as well as the request object which forwards most headers of the initial request.
+// This forwarded request is generated by the http.Client and
+type NextLocationInfo struct {
+	jar             *cookiejar.Jar `json:"-"`
+	location        string         `json:"-"`
+	httpRedirectReq *http.Request  `json:"-"`
+}
+
+type MeasureURLResult struct {
+	CtrlURLMeasurement *CtrlURLMeasurement `json:"-"`
+	redirectedReqs     []*CtrlRequest      `json:"-"`
+	h3Reqs             []*CtrlRequest      `json:"-"`
+}
+
+type MeasureEndpointResult struct {
+	CtrlEndpoint CtrlEndpointMeasurement
+	httpRedirect *NextLocationInfo
+	h3Location   string
+}
+
+func Measure(ctx context.Context, creq *CtrlRequest) (*CtrlResponse, error) {
+	var cresp = &CtrlResponse{URLMeasurements: []*CtrlURLMeasurement{}}
+
+	redirected := make(map[string]bool, 100)
+
+	urlM, err := MeasureURL(ctx, creq, cresp, redirected)
+	if err != nil {
+		return nil, err
+	}
+	cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+
+	n := 0
+	nextRequests := append(urlM.redirectedReqs, urlM.h3Reqs...)
+	for len(nextRequests) > n {
+		req := nextRequests[n]
+		n += 1
+		if len(redirected) == 20 {
+			// stop after 20 redirects
+			break
+		}
+		if _, ok := redirected[req.HTTPRequest]; ok {
+			continue
+		}
+		redirected[req.HTTPRequest] = true
+		urlM, err := MeasureURL(ctx, req, cresp, redirected)
+		if err != nil {
+			return nil, err
+		}
+		cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+		nextRequests = append(nextRequests, urlM.redirectedReqs...)
+	}
```

Comment:
We need to chat about this logic in person: I am not sure I fully understand it.